conversion-measurement/sites/home/shared.browser-setup.js

```diff
@@ -0,0 +1,103 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const MIN_SUPPORTED_BROWSER_VERSION = 105
+const SUPPORTED_USER_AGENTS = ['Google Chrome']
+
+displayBrowserVersionRequirements()
+displayWarningBannerIfSetupIncorrect()
+
+function checkIsBrowserSupported() {
+  if (!navigator.userAgentData) {
+    return false
+  }
+  const supportedAgentsByThisUser = navigator.userAgentData.brands.filter(
+    (item) => SUPPORTED_USER_AGENTS.includes(item.brand)
+  )
+  if (supportedAgentsByThisUser.length === 0) {
+    return false
+  } else {
+    const browserVersion = Number.parseFloat(
+      supportedAgentsByThisUser[0].version
+    )
+    if (browserVersion < MIN_SUPPORTED_BROWSER_VERSION) {
+      return false
+    }
+    return true
+  }
+}
+
+function checkIsFeatureAllowed() {
+  if (!document.featurePolicy) {
+    return false
+  }
+  return document.featurePolicy.allowsFeature('attribution-reporting')
+}
+
+function getBrowserSetupStatus() {
+  const status = {
+    isBrowserSetupCorrect: true,
+    issues: []
+  }
+  const isBrowserVersionSupported = checkIsBrowserSupported()
+  const isFeatureAllowed = checkIsFeatureAllowed()
+  if (!isBrowserVersionSupported || !isFeatureAllowed) {
+    status.isBrowserSetupCorrect = false
+  }
+  if (!isBrowserVersionSupported) {
+    status.issues.push(
+      `Your browser is not supported by this demo. Use ${SUPPORTED_USER_AGENTS.join(
+        ' or '
+      )}, version ${MIN_SUPPORTED_BROWSER_VERSION} or newer.`
+    )
+  }
+  if (!isFeatureAllowed) {
+    status.issues.push(
+      'The Attribution Reporting API is not enabled on this page.'
+    )
+  }
+  return status
+}
+
+function displayBrowserVersionRequirements() {
+  // Display browser version requirements in the demo home page
+  const browserVersionRequirementsEl = document.getElementById(
+    'browserVersionRequirementsEl'
+  )
+  if (browserVersionRequirementsEl) {
+    browserVersionRequirementsEl.innerText = `Use ${SUPPORTED_USER_AGENTS.join(
+      ' or '
+    )}, version ${MIN_SUPPORTED_BROWSER_VERSION} or newer.`
+  }
+}
+
+function displayWarningBannerIfSetupIncorrect() {
+  // Check browser setup and display banner if needed
+  const { isBrowserSetupCorrect, issues } = getBrowserSetupStatus()
+
+  if (!isBrowserSetupCorrect) {
+    const banner = document.getElementById('notSupportedBanner')
+    if (banner.style.display === 'none') {
+      banner.style.display = 'block'
+      banner.innerHTML += `Your setup seems incorrect.<br>`
```

Comment:
I'm not sure we want use `innerHTML +=` here, as if the function is called more than once, the `Your setup seems incorrect` message will be appended multiple times.

Instead, we may want to include an element whose content we can completely replace each time this function is called, e.g. a `<div>` with a known ID.

Reply:
SGTM, updated the code accordingly:

https://github.com/GoogleChromeLabs/trust-safety-demo/pull/34/files#diff-872cdba4a67ecaa9521b96521beabb827d4eafb269f577a3853b33defa709d3eR12

https://github.com/GoogleChromeLabs/trust-safety-demo/pull/34/files#diff-95956c9070e63fd7b9c9d2b916efbc158993f6e231f4ff628fc72a85b99fca3bR91